dlib/dnn/loss_abstract.h

```diff
@@ -1852,6 +1852,191 @@ namespace dlib
+        /*!
+            ensures
+                - #get_options() == options_
+        !*/
+
+        const yolo_options& get_options (
+        ) const;
+        /*!
+            ensures
+                - returns the options object that defines the general behavior of this loss layer.
+        !*/
+
+        template <
+            typename SUB_TYPE,
+            typename label_iterator
+            >
+        void to_label (
+            const tensor& input_tensor,
+            const SUB_TYPE& sub,
+            label_iterator iter,
+            double adjust_threshold = 0.25
+        ) const;
+        /*!
+            This function has the same interface as EXAMPLE_LOSS_LAYER_::to_label() except
+            it has the additional calling requirements that:
+                - layer<TAG_TYPE>(sub).get_output().k() == options.anchors.at(tag_id<TAG_TYPE>::id).size() * (5 + options.labels.size());
+                - sub.get_output().num_samples() == input_tensor.num_samples()
+                - sub.sample_expansion_factor() == 1
+            Also, the output labels are std::vectors of yolo_rects where, for each yolo_rect R,
+            we have the following interpretations:
+                - R.rect == the location of an object in the image.
+                - R.detection_confidence == the score for the object, between 0 and 1.  Only
+                  objects with a detection_confidence > adjust_threshold are output.  So if
+                  you want to output more objects (that are also of less confidence) you
+                  can call to_label() with a smaller value of adjust_threshold.
+                - R.label == the label of the detected object.
+                - R.labels == a std::vector<double, std::string> containing all the confidence values
+                  and labels that have a detection score > adjust_threshold, since this loss allows
+                  for multi-label outputs.  Note that the following is true:
+                      - R.labels[0].first == R.detection_confidence
+                      - R.labels[0].second == R.label
+                - R.ignore == false (this value is unused by to_label()).
+        !*/
+
+        template <
+            typename const_label_iterator,
+            typename SUBNET
+            >
+        double compute_loss_value_and_gradient (
+            const tensor& input_tensor,
+            const_label_iterator truth, 
+            SUBNET& sub
+        ) const;
+        /*!
+            This function has the same interface as EXAMPLE_LOSS_LAYER_::compute_loss_value_and_gradient() 
+            except it has the additional calling requirements that: 
+                - layer<TAG_TYPE>(sub).get_output().k() == options.anchors.at(tag_id<TAG_TYPE>::id).size() * (5 + options.labels.size());
+                - sub.get_output().num_samples() == input_tensor.num_samples()
+                - sub.sample_expansion_factor() == 1
+            Also, the loss value returned corresponds to the L2 loss.
```

Comment:
Yes, I have mixed feelings about this. As you might have seen in the code, we're just computing the difference between the output of the network and the target (which will act as the gradient) and then squaring those to display the loss...

``` c++
loss += length_squared(rowm(mat(grad), n));
```
Even for the objectness and classes (I know they are the outputs of sigmoid functions) but it just seems like the L2 loss to me: the sum of the squared differences between the outputs and the targets.

Reply:
Let me look at the code again, maybe I'm not interpreting what you are saying correctly.  

Like are you saying that the loss value coming out here isn't the loss being optimized?  Rather that it's only part of it?  The loss value coming out should be the value of the objective function the loss is optimizing.  Like the gradient of this value is what defines the gradient (or it should).   Otherwise it's pretty confusing. And stuff like testing when the loss has stopped improving isn't meaningful if the loss value isn't the actual loss being optimized.

Reply:
Oh right, I see.  The `loss` is being set to the norm of the loss's gradient.  It should be set to the loss value though.  Can you adjust it so `loss` is the loss value?

Reply:
I mean, IDK, maybe it doesn't matter  It's technically wrong though as the value being output is not the loss value. It's the square of the norm of the gradient vector of that loss, which is in general a different value.  It's probably going to end up making training run fine.  But anyone looking at that value is going to be confused when it's some other number.

Reply:
So, do you mean computing the (multi)binary log loss for the objectness and the classification, and the L2 loss for the bounding box regression?

I did it this way because it's how it's done in the [official YOLO implementation](https://github.com/pjreddie/darknet/blob/masterhttps://github.com/pjreddie/darknet/blob/master/src/yolo_layer.c#L238):

```c
*(l.cost) = pow(mag_array(l.delta, l.outputs * l.batch), 2);
```
Where `l.delta` contains the gradients, and `mag_array` is the sum of the squared differences.

Reply:
But the official YOLO implementation doesn't use that value to make any decision, just for displaying purposes.

Reply:
Right, the loss value is really just for display and for doing stuff like checking for convergence.

It's just funny because it's wrong according to the interface contract. 

So if you want to keep using this value instead of the loss I think that's fine. Update the docs for the loss layer to call out that the loss value isn't the loss but the squared gradient norm though. Do the logged loss value will be that instead of the loss. 

And you have found that the convergence testing stuff all works fine with this as the loss? Seems like it would be fine.

Reply:
I've been trying this loss for about 1 month, and it seems to work fine, but maybe it's not working at its full potential.
However, if I introduce that change, I will need to test it again (it might take some days, maybe a week)

Reply:
Yeah. Don't worry about it. Do that in another PR if you want :)

For this one update the docs and I'll hit the merge button :)

Reply:
ok :)

Reply:
Done